crates/ledger/src/lib.rs

```diff
@@ -92,88 +165,31 @@ impl<T: Store> gasket::framework::Worker<Stage<T>> for Worker {
-                    target: EVENT_TARGET,
-                    "forward",
-                    header.height = tracing::field::Empty,
-                    header.slot = tracing::field::Empty,
-                    header.hash = tracing::field::Empty,
-                    stable.epoch = tracing::field::Empty,
-                    tip.epoch = tracing::field::Empty,
-                    tip.relative_slot = tracing::field::Empty,
-                )
-                .entered();
-
-                let span_parse_block = debug_span!(
-                    target: EVENT_TARGET,
-                    parent: &span_forward,
-                    "parse_block",
-                    point = point.slot_or_default(),
-                    block.size = raw_block.len(),
-                )
-                .entered();
-
-                let (block_header_hash, block) = parse_block(&raw_block[..]);
-
-                span_parse_block.exit();
-
-                span_forward.record("header.height", block.header.header_body.block_number);
-                span_forward.record("header.slot", block.header.header_body.slot);
-                span_forward.record("header.hash", hex::encode(block_header_hash));
-
-                let mut state = stage.state.lock().await;
-
-                state.forward(&span_forward, point, block).map_err(|e| {
-                    error!(target: EVENT_TARGET, error = ?e, "forward.failed");
-                    WorkerError::Panic
-                })?;
-
-                span_forward.exit();
+        let result = match unit {
+            ValidateBlockEvent::Validated(point, raw_block) => {
+                stage
+                    .validate_block(point.clone(), raw_block.to_vec())
+                    .await
             }
 
-            ValidateHeaderEvent::Rollback(point) => {
-                let span_backward = debug_span!(
-                    target: EVENT_TARGET,
-                    "backward",
-                    point.slot = point.slot_or_default(),
-                    point.hash = tracing::field::Empty,
-                )
-                .entered();
-
-                if let Point::Specific(_, header_hash) = point {
-                    span_backward.record("point.hash", hex::encode(header_hash));
-                }
-
-                let mut state = stage.state.lock().await;
-
-                if let Err(e) = state.backward(point) {
-                    match e {
-                        BackwardErr::UnknownRollbackPoint(_) => {
-                            warn!(target: EVENT_TARGET, "rollback.unknown_point")
-                        }
-                    }
-                }
-
-                span_backward.exit();
-            }
-        }
+            ValidateBlockEvent::Rollback(point) => stage.rollback_to(point.clone()).await,
+        };
 
-        Ok(())
+        Ok(stage.downstream.send(result.into()).await.or_panic()?)
     }
 }
 
+#[instrument(skip(bytes), fields(block.size = bytes.len()))]
```

Comment:
```suggestion
#[instrument(level = Level::DEBUG, skip(bytes), fields(block.size = bytes.len()))]
```

Reply:
Note I was about to do the same for `validate_block` but we pass the `span` to called functions so it's more involved, I will do that in another PR perhaps?